### PR TITLE
feat(): add ManagedIdentity in Azure Blob Storage

### PR DESCRIPTION
**Pull request by sandy2008**
- Support Using ManagedIdentity in Azure Blob Storage

### CODE REVIEW

**Comment by hanayashiki** [comment]
wow, this looks super useful

File: pkg/storage/chunk/azure/blob_storage_client.go
Comment (DylanGuedes) [inline pkg/storage/chunk/azure/blob_storage_client.go:240]:
```suggestion
	if !b.cfg.UseManagedIdentity {
```

File: pkg/storage/chunk/azure/blob_storage_client.go
Comment (DylanGuedes) [inline pkg/storage/chunk/azure/blob_storage_client.go:246]: This is the same option from line 273. Consider defining it before the if clause and reusing it.

File: pkg/storage/chunk/azure/blob_storage_client.go
Comment (DylanGuedes) [inline pkg/storage/chunk/azure/blob_storage_client.go:283]: nit
```suggestion
```

File: pkg/storage/chunk/azure/blob_storage_client.go
Comment (DylanGuedes) [inline pkg/storage/chunk/azure/blob_storage_client.go:320]:
```suggestion
```

File: pkg/storage/chunk/azure/blob_storage_client.go
Comment (DylanGuedes) [inline pkg/storage/chunk/azure/blob_storage_client.go:321]:
```suggestion
```

File: pkg/storage/chunk/azure/blob_storage_client.go
Comment (DylanGuedes) [inline pkg/storage/chunk/azure/blob_storage_client.go:324]:
```suggestion
	spt, err := adal.NewServicePrincipalTokenFromMSI(msiEndpoint, "https://storage.azure.com/")
```

File: pkg/storage/chunk/azure/blob_storage_client.go
Comment (DylanGuedes) [inline pkg/storage/chunk/azure/blob_storage_client.go:267]: nit: since the if clause ends with a return, you can safely move the else code to outside of it.
```suggestion
	}
```

**Review by DylanGuedes — COMMENTED** [review]
a few nits

**Comment by gai6948** [comment]
looking forward to this feature, want to use that on our AKS clusters

**Comment by cyriltovena** [comment]
Can you run `make check-mod` and push the result ? Do you nee help finishing this ?

**Comment by sandy2008** [comment]
> Can you run `make check-mod` and push the result ? Do you nee help finishing this ?

Sorry I was little busy recently, I am wrapping it up!
Thank you so much for your kind comments!

File: pkg/storage/chunk/azure/blob_storage_client.go
Comment (sandy2008) [reply pkg/storage/chunk/azure/blob_storage_client.go:247]: thx, resolved.

File: pkg/storage/chunk/azure/blob_storage_client.go
Comment (sandy2008) [reply pkg/storage/chunk/azure/blob_storage_client.go:289]: thx, resolved.

File: pkg/storage/chunk/azure/blob_storage_client.go
Comment (sandy2008) [reply pkg/storage/chunk/azure/blob_storage_client.go:327]: thx, resolved.

**Comment by sandy2008** [comment]
> a few nits

Thx, I fixed all of them.

**Comment by sandy2008** [comment]
@DylanGuedes @cyriltovena Please kindly review it again when you got time.

**Comment by hanayashiki** [comment]
can‘t wait for this to be merged 🙏

**Review by DylanGuedes — APPROVED** [review]
just a nit but LGTM! (I'm not a maintainer, though)

File: pkg/storage/chunk/azure/blob_storage_client.go
Comment (cyriltovena) [inline pkg/storage/chunk/azure/blob_storage_client.go:247]:
```suggestion
	// defining the Azure Pipeline Options
```

File: pkg/storage/chunk/azure/blob_storage_client.go
Comment (cyriltovena) [inline pkg/storage/chunk/azure/blob_storage_client.go:285]: I think the only change required was this credentials object, all other options should be the same wether you used managed identitfy or not.

I would remove this return, use the same opts and credentials object, move the if down and override the default credentials when required.

Comment (sandy2008) [reply pkg/storage/chunk/azure/blob_storage_client.go:285]: Thank you, do you want me to make the change & push it?

Comment (cyriltovena) [reply pkg/storage/chunk/azure/blob_storage_client.go:285]: Yeah I think it's just about moving the code around to ensure the client is configure for all type of credentials.

Comment (sandy2008) [reply pkg/storage/chunk/azure/blob_storage_client.go:286]: OK, I pushed another commit, could you take a look if I am correct..? Thank you so much!!

**Review by cyriltovena — APPROVED** [review]
LGTM

**Pull request merged**
